kratos/processes/vtk_output_process.cpp

```diff
@@ -0,0 +1,734 @@
+
+void VtkOutputProcess::ExecuteInitialize()
+{}
+
+void VtkOutputProcess::ExecuteBeforeSolutionLoop()
+{}
+
+void VtkOutputProcess::ExecuteInitializeSolutionStep()
+{}
+
+void VtkOutputProcess::ExecuteFinalizeSolutionStep()
+{
+    PrintOutput();
+}
+
+void VtkOutputProcess::ExecuteBeforeOutputStep()
+{}
+
+void VtkOutputProcess::ExecuteAfterOutputStep()
+{}
+
+void VtkOutputProcess::ExecuteFinalize()
+{}
+
+int VtkOutputProcess::Check()
+{return  0;}
+
+
+
+void VtkOutputProcess::CreateMapFromKratosIdToVTKId(ModelPart &model_part)
+{
+    int vtk_id = 0;
+
+    for (ModelPart::NodeIterator node_i = model_part.NodesBegin(); node_i != model_part.NodesEnd(); ++node_i)
+    {
+        int KratosId = node_i->Id();
+        mKratosIdToVtkId[KratosId] = vtk_id;
+        vtk_id++;
+    }
+}
+
+unsigned int VtkOutputProcess::DetermineVtkCellListSize(ModelPart &model_part)
+{
+    unsigned int vtk_cell_list_size = 0;
+
+    for (ModelPart::ElementIterator elem_i = model_part.ElementsBegin(); elem_i != model_part.ElementsEnd(); ++elem_i)
+    {
+        vtk_cell_list_size++;
+        vtk_cell_list_size += elem_i->GetGeometry().size();
+    }
+
+    for (ModelPart::ConditionIterator condition_i = model_part.ConditionsBegin(); condition_i != model_part.ConditionsEnd(); ++condition_i)
+    {
+        vtk_cell_list_size++;
+        vtk_cell_list_size += condition_i->GetGeometry().size();
+    }
+
+    return vtk_cell_list_size;
+}
+
+void VtkOutputProcess::Initialize(ModelPart &model_part)
+{
+    CreateMapFromKratosIdToVTKId(model_part);
+    mVtkCellListSize = DetermineVtkCellListSize(model_part);
+}
+
+void VtkOutputProcess::WriteHeader(ModelPart &model_part)
+{
+    std::string outputFileName = GetOutputFileName(model_part);
+    std::ofstream outputFile;
+
+    outputFile.open(outputFileName, std::ios::out | std::ios::binary | std::ios::trunc);
+    outputFile << "# vtk DataFile Version 4.0"
+               << "\n";
+    outputFile << "vtk output"
+               << "\n";
+    outputFile << "ASCII"
+               << "\n";
+    outputFile << "DATASET UNSTRUCTURED_GRID"
+               << "\n";
+    outputFile.close();
+}
+
+void VtkOutputProcess::WriteMesh(ModelPart &model_part)
+{
+    WriteNodes(model_part);
+    WriteConditionsAndElements(model_part);
+    WriteConditionAndElementTypes(model_part);
+}
+
+void VtkOutputProcess::WriteNodes(ModelPart &model_part)
+{
+    std::string outputFileName = GetOutputFileName(model_part);
+    std::ofstream outputFile;
+    outputFile.open(outputFileName, std::ios::out | std::ios::app | std::ios::binary);
+    outputFile << std::scientific;
+    outputFile << std::setprecision(mDefaultPrecision);
+
+    // write nodes header
+    outputFile << "POINTS " << model_part.NumberOfNodes() << " float"
+               << "\n";
+
+    // write nodes
+    for (ModelPart::NodeIterator node_i = model_part.NodesBegin(); node_i != model_part.NodesEnd(); ++node_i)
+    {
+        double x_coordinate = node_i->X();
```

Comment:
Not very efficient. Do const array& coor = node->Coordinates(), so you only access once and is const

Reply:
Please use the coordinates for a better performance